src/Waives.Pipelines/WorkPool.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Waives.Pipelines
+{
+    internal class WorkPool : IDisposable
+    {
+        private readonly int _maxConcurrency;
+        private readonly SemaphoreSlim _semaphore;
+        private readonly ConcurrentDictionary<Task, bool> _tasks = new ConcurrentDictionary<Task, bool>();
+
+        public WorkPool(int maxConcurrency)
+        {
+            _maxConcurrency = maxConcurrency;
+            _semaphore = new SemaphoreSlim(maxConcurrency, maxConcurrency);
+        }
+
+        /// <summary>
+        /// Post accepts an async function to run concurrently, up to the maximum level
+        /// specified in the constructor. If the WorkPool is already running at the maximum
+        /// level of concurrency, Post blocks until one of the in-flight Tasks has completed.
+        /// </summary>
+        /// <param name="action"></param>
+        public void Post(Func<Task> action)
```

Comment:
💪❓
```suggestion
        public void Schedule(Func<Task> action)
```

Reply:
> Please could you clarify why Polly's Bulkhead isolation policy didn't meet our requirements? I understand the difference in behaviour you described in the PR, but I don't understand _why_ we want to halt the entire pipeline as opposed to buffering it. Both seem to be valid approaches as far as I can tell (and there might be some benefits to the latter approach).

The trouble is that with the bulkhead, we need to specify a queue length. Given we have no idea how many documents we may be processing, there's no way of knowing what the 'right' queue length is. If it's too short, polly will throw an exception when we try to buffer the next document. Too long, and we end up with lots of `Document` instances in memory, but even worse is that if an error occurs before they all get processed, they will have been consumed (and assumed processed) from the source observable. In essence, they'd be lost.

I think we need to block if we reach our concurrency limit, and in fact that's the fundamental reason for this PR - the now-removed `RateLimiter` would allow all source documents to be received by our processing observable, potentially keeping 100s or 1000s (or 1000000s) of `Document`s in memory before they could be processed. There would be no back-pressure - the source observable could continue issuing source documents as whatever rate it pleased, regardless of how fast they were being processed. Making `WorkPool` block applies that back-pressure.

> 
> I also have a small concern that the tests make show intermittent failures. Please can you verify this is not the case?

Certainly - I haven't seen any intermittent failures...!